doc/dev/adr/0023-property-support.md

```diff
@@ -0,0 +1,340 @@
+| --- | --- | --- | --- | --- |
+| `propertyTopic` | optional | *string* | slash-separated sequence of character-restricted labels and/or brace-enclosed tokens | MQTT topic pattern on which a read or write request is published. |
+
+The DTDL Mqtt extension places no restrictions &mdash; other than basic syntactical constraints &mdash; on the set of tokens used in MQTT topic patterns.
+Topic tokens recognized by the ProtocolCompiler are defined in the [topic-structure](https://github.com/Azure/iot-operations-sdks/blob/main/doc/reference/topic-structure.md) document.
+The sets of tokens differ between RPC and Telemetry, and the following set of tokens is hereby defined for Property:
+
+| Topic token | Description | Required |
+| --- | --- | --- |
+| `{modelId}` | The identifier of the service model, which is the full DTMI of the Interface | optional |
+| `{maintainerId}` | Identifier of the maintainer, by default the MQTT client ID | optional |
+| `{consumerClientId}` | The MQTT client ID of the endpoint that requests to read or write a Property | optional |
+| `{propertyName}` | The name of the Property | optional |
+| `{action}` | "read" or "write" | optional |
+
+>[!NOTE]
+> The string value of the `propertyTopic` property is not required to contain the `{propertyName}` token, but Property communication differs depending on whether this token is present:
+>
+> * If the `propertyTopic` property contains a `{propertyName}` token, each Property is assigned a separate MQTT pub/sub topic, and each is read or written separately.
+> * If the `propertyTopic` property does not contain a `{propertyName}` token, all Properties in the Interface are grouped into a collection that is read or written with a combined payload. This does not require updating all Properties with each write operation, but it provides the option of writing multiple Properties in a single consistent update.
+
+>[!NOTE]
+> The action token is optional because a model might have no writable properties, in which case all requests are implicitly read requests.
+
+#### Basic sample model with separate Property topics
+
+The following sample model defines two Properties, one read-only and one writable.
+The property topic pattern contains the tokens "{propertyName}" and "{action}".
+
+```json
+{
+  "@context": [ "dtmi:dtdl:context;4", "dtmi:dtdl:extension:mqtt;4" ],
+  "@id": "dtmi:propertySketch:PropertySketch;1",
+  "@type": [ "Interface", "Mqtt" ],
+  "payloadFormat": "Json/ecma/404",
+  "propertyTopic": "sample/property/{propertyName}/{action}",
+  "contents": [
+    {
+      "@type": "Property",
+      "name": "Foo",
+      "schema": "integer"
+    },
+    {
+      "@type": "Property",
+      "name": "Bar",
+      "writable": true,
+      "schema": "string"
+    }
+  ]
+}
+```
+
+This model generates three Commands:
+
+| Command name | MQTT topic | Example request | Example response |
+| --- | --- | --- | --- |
+| ReadFoo | "sample/property/Foo/read" | | { "Foo": 33 } |
```

Comment:
Do we need the property name in the read/write body when it is specified in the command/topic? That seems redundant (though I could see an argument for consistency between this and the grouped format).

Reply:
We don't need the name per se, but we do need *some* field in the body.  Due to restrictions of some formats (AVRO and Protobuf, specifically), the payload must be an object, not a primitive type like a string or number.  So, the minimum requirement is an object with a single field, and that field needs a name.  For consistency, we use the same name as the aggregate format uses.  This also aligns with the way Telemetry is defined.